provider/consul/consul_catalog.go

```diff
@@ -93,33 +95,117 @@ func (p *CatalogProvider) watchServices(stopCh <-chan struct{}) <-chan map[strin
 			default:
 			}
 
-			data, catalogMeta, err := catalog.Services(catalogOptions)
-			if err != nil {
-				log.WithError(err).Error("Failed to list services")
-				return
-			}
-
 			// Listening to changes that leads to `passing` state or degrades from it.
-			// The call is used just as a trigger for further actions
-			// (intentionally there is no interest in the received data).
-			_, healthMeta, err := health.State("passing", healthOptions)
+			data, meta, err := health.State("passing", options)
 			if err != nil {
 				log.WithError(err).Error("Failed to retrieve health checks")
 				return
 			}
 
 			// If LastIndex didn't change then it means `Get` returned
 			// because of the WaitTime and the key didn't changed.
-			sameServiceAmount := catalogOptions.WaitIndex == catalogMeta.LastIndex
-			sameServiceHealth := lastHealthIndex == healthMeta.LastIndex
-			if sameServiceAmount && sameServiceHealth {
+			if options.WaitIndex == meta.LastIndex {
+				continue
+			}
+
+			options.WaitIndex = meta.LastIndex
+
+			if data != nil {
+				// Exclude node level checks
+				checks := fun.Filter(func(check *api.HealthCheck) bool {
+					return check.ServiceName != ""
+				}, data).([]*api.HealthCheck)
+
+				// A critical note is that the return of a blocking request is no guarantee of a change.
+				// It is possible that there was an idempotent write that does not affect the result of the query.
+				services := fun.Set(fun.Map(func(check *api.HealthCheck) string {
+					return check.ServiceName
```

Comment:
I was just filtering only services with real names without `consul` agents. But now I adjusted the code to work only with Catalog Services response. Should be more cleaner approach now.